cpp/src/round_robin/round_robin.cu

```diff
@@ -0,0 +1,130 @@
+  CUDF_EXPECTS( start_partition < num_partitions, "Incorrect start_partition index. Must be less than number of partitions." );
+  
+  auto n_pmax = nrows % num_partitions;//# partitions of max size
+  cudf::size_type max_p_size = std::ceil( static_cast<double>(nrows) / static_cast<double>(num_partitions));// max size of partitions
+  
+  auto pmm = n_pmax * max_p_size;
+  auto dnpmax = num_partitions - n_pmax;
+
+  //delta is the number of positions to rotate right
+  //the original range [0,1,...,n-1]
+  //and is calculated by accumulating the first
+  //`start_partition` partition sizes from the end;
+  //i.e.,
+  //the partition sizes array (of size p) being:
+  //[m,m,...,m,(m-1),...,(m-1)]
+  //(with n_pmax sizes `m` at the beginning;
+  //and (p-n_pmax) sizes `(m-1)` at the end)
+  //we accumulate the 1st `start_partition` entries from the end:
+  //
+  auto delta = (start_partition > dnpmax?
+                dnpmax*(max_p_size-1) + (start_partition - dnpmax)*max_p_size :
+                start_partition*(max_p_size-1));
+
+  
+  auto iter_begin =
+    thrust::make_transform_iterator(thrust::make_counting_iterator<cudf::size_type>(0),
+                                    [nrows, num_partitions, max_p_size, n_pmax, pmm, delta] __device__ (auto index0){
+                                      auto indx = (index0 + nrows - delta) % nrows;//rotate original index right by delta positions
+                                      auto ipj = (indx <= pmm ? indx % max_p_size: (indx - pmm) % (max_p_size-1) );
+                                      auto pij = (indx <= pmm ? indx / max_p_size: n_pmax + (indx - pmm) / (max_p_size-1) );
+                                      return num_partitions * ipj + pij;
+                                    });
+
+  auto uniq_tbl = cudf::experimental::detail::gather(input,
+                                                     iter_begin, iter_begin + nrows,
+                                                     false, false, false,
+                                                     mr,
+                                                     stream);
+
+  auto ret_pair =
+    std::make_pair(std::move(uniq_tbl), std::vector<cudf::size_type>(num_partitions));
+  rmm::device_vector<cudf::size_type> d_partition_offsets(num_partitions, cudf::size_type{0});
+
+  auto rotated_iter_begin =
+    thrust::make_transform_iterator(thrust::make_counting_iterator<cudf::size_type>(0),
+                                    //(\indx -> if (indx + np - sp) `mod` np < pmax then m else (m-1) )
+                                    [num_partitions, start_partition, max_p_size, n_pmax] __device__ (auto indx){
+                                      return ((indx + num_partitions - start_partition) % num_partitions < n_pmax? max_p_size : max_p_size-1);
+                                    });
+
+  auto exec = rmm::exec_policy(stream);
+  thrust::exclusive_scan(exec->on(stream),
```

Comment:
`num_partitions` is usually going to be relatively small (<1,000), as such, it's probably more expensive to do this on the device. Instead, do it on the host directly into the `std::vector` and avoid the memcpy.

Reply:
Hmm. I'm getting failures, using:
```
thrust::exclusive_scan(thrust::host,
                         rotated_iter_begin, rotated_iter_begin + num_partitions,
                         ret_pair.second.begin());```

The `exclusive_scan` seems to fail on host (I get [0...0] as partition offsets).

Reply:
Oops, I think I know why: the transform iterator uses a device lambda.

Reply:
Done.